### PR TITLE
Support to add resource via class parameters. Add manage_hostgroup_fo…

### PR DESCRIPTION
**Pull request by MaxFedotov**
Add ability to configure resources through class parameters using variables (`$mysql_servers`, `$mysql_users`, `$mysql_hostgroups`, `$mysql_rules`, `$schedulers`)
New param `manage_hostgroup_for_servers`. This defines which provider will be called from mysql_servers. If true - default (proxy_mysql_server). If false - proxy_mysql_server_no_hostgroup, which don't look on hostgroup field of mysql_server. It is needed in order to allow configuration of this parameter by some external service  - consul, script, etc. This can close https://github.com/voxpupuli/puppet-proxysql/issues/38

### CODE REVIEW

File: examples/init.pp [hunk 1]
```diff
@@ -1,4 +1,34 @@
 # lint:ignore:80chars
+# lint:ignore:2sp_soft_tabs
+# variant 1
+
+class { '::proxysql':
```
Comment (bastelfreak) [inline examples/init.pp]: could you move the example into the README.md? We noticed that people do not really look into the examples folder :(

File: manifests/configure.pp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      $server.each |$k,$v| {
+        $hostname = $k
+        $port = $server[$k][port] ? { undef   => 3306,
+                                      default => $server[$k][port], }
+        $hostgroup_id = $server[$k][hostgroup_id]
+
+        if $proxysql::manage_hostgroup_for_servers {
+          proxy_mysql_server { "${hostname}:${port}-${hostgroup_id}":
+            hostname => $hostname,
+            *        => $server[$k],
```
Comment (bastelfreak) [inline manifests/configure.pp]: nice, I rarely see this syntax <3

File: manifests/init.pp [hunk 3]
```diff
@@ -182,6 +207,7 @@
   -> class { '::proxysql::config':}
   -> class { '::proxysql::service':}
   -> class { '::proxysql::admin_credentials':}
+  -> class { '::proxysql::configure':}
```
Comment (bastelfreak) [inline manifests/init.pp]: please do not add the leading `::`

File: types/rule.pp [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# lint:ignore:2sp_soft_tabs
+type Proxysql::Rule = Array[Hash[String, Struct[{ rule_id                         => Integer,
```
Comment (bastelfreak) [inline types/rule.pp]: wow :1st_place_medal:

Comment (bastelfreak) [reply types/rule.pp]: https://twitter.com/BastelsBlog/status/1035652506813169671

**Comment by bastelfreak** [comment]
Wow, what a massive amout of work, thanks for this! As in the other PRs, please check the used email address.

File: examples/init.pp (hunk 1 above)
Comment (MaxFedotov) [reply examples/init.pp]: done

**Comment by MaxFedotov** [comment]
All review notes were merges

File: README.markdown [hunk 5]
```diff
@@ -59,6 +59,111 @@ You can customize options such as (but not limited to) `listen_port`, `admin_pas
   }
 ```
 
+You can configure users\hostgroups\rules\schedulers using class parameters
+```
```
Comment (bastelfreak) [inline README.markdown]: please enable the Puppet syntax highlighting:
```
```puppet
```

Comment (MaxFedotov) [reply README.markdown]: done

**Comment by mcrauwel** [comment]
@MaxFedotov can you rebase on the current master?

**Comment by MaxFedotov** [comment]
done

**Comment by mcrauwel** [comment]
let's continue with this now, of course on every PR we merge, another rebase will be required...

**Comment by MaxFedotov** [comment]
done

**Pull request merged**